text/metrics/0146-metrics-prototype-scenarios.md

```diff
@@ -0,0 +1,220 @@
+# Scenarios for Metrics API/SDK Prototyping
+
+With the stable release of the tracing specification, the OpenTelemetry
+community is willing to spend more energy on metrics API/SDK. The goal is to get
+the metrics API/SDK specification to
+[`Experimental`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#experimental)
+state by end of 5/2021, and make it
+[`Stable`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#stable)
+before end of 2021:
+
+* By end of 5/2021, we should have a good confidence that we can recommend
+  language client owners to work on metrics preview release. This means starting
+  from 7/1/2021 the specification should not have major surprise or big changes
+  which would thrust the language client maintainers.
+
+* By end of 9/2021, we should mark the metrics API/SDK specification as
+  [`Feature-freeze`](https://github.com/open-telemetry/opentelemetry-specification/blob/1afab39e5658f807315abf2f3256809293bfd421/specification/document-status.md#feature-freeze),
+  and focusing on bug fixing or editorial changes.
+
+* By end of 2021, we want to have a stable release of metrics API/SDK
+  specification, with multiple language SIGs providing RC (release candidate) or
+  [stable](https://github.com/open-telemetry/opentelemetry-specification/blob/9047c91412d3d4b7f28b0f7346d8c5034b509849/specification/versioning-and-stability.md#stable)
+  clients.
+
+In this document, we will focus on two scenarios that we use for prototyping
+metrics API/SDK. The goal is to have two scenarios which clearly capture the
+major requirements, so we can work with language client SIGs to prototype,
+gather the learnings, determine the scopes and stages. Later the scenarios can
```

Comment:
Could you provide an example of one of these?

Reply:
There are 6 instrumented in current [API](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#the-instruments). Anything from that list which is not used for the above 2 scenarios.

Reply:
Another example I am anticipating are the batching APIs. My cursory review of some other pre-existing metric APIs suggested that atomic modification of multiple metrics wasn't a common feature and nothing seemed apparent to me in these example scenarios that required batching to exist.

Reply:
The draft OTel APi included a number of things that were supposedly demanded for performance. One of those was a batch API for recording metrics, but while we agreed on that there were many unanswered questions about the bigger picture--particularly because there is a conflicting request for per-instrument collection intervals. Nothing that can't be solved, just growing scope and complexity.

> Could you provide an example of one of these?

There are no Gauge use-cases in these scenarios. Although the 6 instruments in the draft API cover most scenarios, I've become convinced that users will continue to ask for a synchronous form of Gauge, which makes at least 7 use-cases, but even if we are convinced that all valid use-cases can be mapped into 7 fundamental instruments, it doesn't mean users want those APIs.

For the current scenario 2, there's a relatively heavy-weight instrumentation called for in a high-value piece of instrumentation. You are likely to invest in more than one line of code for a timing measurement in that case.

OTOH, in less high-value code, I'm likely to use a one-liner for measuring a duration metric. Here's an example OTel-Go `ValueRecorder` wrapper that emits a timing measurement using a one-line expression like

```
func myConnect() (retErr error) {
  defer connectDuration.Start(ctx).Stop(&retErr)
  // ... 
}
```

It's not clear that 7 scenarios is enough, but I think starting with 2 is great and avoiding Gauges to start is :100:.

Reply:
> It's not clear that 7 scenarios is enough, but I think starting with 2 is great and avoiding Gauges to start is 💯.

+100 💯

Reply:
We will need to take the step by step approach.

For things that are currently in the draft spec, but not covered/used by the scenarios here, we should table them for now and only revisit them after we covered the basics.

Please refer to the next section - "we will need to have an incremental approach and divide the work into multiple stages", it should clarify the approach we're taking.

Reply:
Sounds great, thanks for the clarification : )